src/he/pmtud.c

```diff
@@ -0,0 +1,287 @@
+        conn->pmtud_probing_size -= PMTUD_PROBE_SMALL_STEP;
+        return he_internal_pmtud_search_completed(conn);
+      }
+      break;
+    case HE_PMTUD_STATE_SEARCH_COMPLETE:
+      // Black hole detected
+      return he_internal_pmtud_blackhole_detected(conn);
+    default:
+      // Do nothing in DISABLED / ERROR states
+      break;
+  }
+  return HE_SUCCESS;
+}
+
+he_return_code_t he_internal_pmtud_start_base_probing(he_conn_t *conn) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+
+  // Check current state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_DISABLED:
+    case HE_PMTUD_STATE_SEARCHING:
+    case HE_PMTUD_STATE_SEARCH_COMPLETE:
+      // Valid states
+      break;
+    default:
+      // Invalid states
+      return HE_ERR_INVALID_CONN_STATE;
+  }
+
+  // Change state
+  he_internal_pmtud_change_state(conn, HE_PMTUD_STATE_BASE);
+
+  // Initialize PMTUD internal state
+  conn->pmtud_base = MAX(MIN_PLPMTU, conn->effective_pmtu);
+  conn->pmtud_probe_count = 0;
+
+  // Start probing base mtu
+  return he_internal_pmtud_send_probe(conn, conn->pmtud_base);
+}
+
+he_return_code_t he_internal_pmtud_base_confirmed(he_conn_t *conn) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+  // Check current state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_BASE:
+    case HE_PMTUD_STATE_ERROR:
+      // Valid states
+      break;
+    default:
+      // Invalid states
+      return HE_ERR_INVALID_CONN_STATE;
+  }
+
+  // Change to Searching state
+  he_internal_pmtud_change_state(conn, HE_PMTUD_STATE_SEARCHING);
+
+  // Start searching
+  conn->pmtud_probe_count = 0;
+  conn->pmtud_is_using_big_step = true;
+  conn->pmtud_probing_size = conn->pmtud_base + PMTUD_PROBE_BIG_STEP;
+
+  return he_internal_pmtud_send_probe(conn, conn->pmtud_probing_size);
+}
+
+he_return_code_t he_internal_pmtud_confirm_base_failed(he_conn_t *conn) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+  // Check current state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_BASE:
+      // Valid states
+      break;
+    default:
+      // Invalid states
+      return HE_ERR_INVALID_CONN_STATE;
+  }
+
+  // Change state
+  he_internal_pmtud_change_state(conn, HE_PMTUD_STATE_ERROR);
+
+  // TODO: recheck base pmtu periodically
+  conn->pmtud_probe_count = 0;
+
+  return HE_SUCCESS;
+}
+
+he_return_code_t he_internal_pmtud_search_completed(he_conn_t *conn) {
+  if(!conn) {
+    return HE_ERR_NULL_POINTER;
+  }
+  // Check current state
+  switch(conn->pmtud_state) {
+    case HE_PMTUD_STATE_SEARCHING:
+      // Valid states
+      break;
+    default:
+      // Invalid states
+      return HE_ERR_INVALID_CONN_STATE;
+  }
+
+  // Set the effective pmtu
+  conn->effective_pmtu = MIN(conn->pmtud_probing_size, MAX_PLPMTU);
+
+  // Change state
+  he_internal_pmtud_change_state(conn, HE_PMTUD_STATE_SEARCH_COMPLETE);
+
+  // TODO: stay in this state and check for black hole regularly
```

Comment:
If we are going to stay in this state, we need to disable the timer_cb (through `conn->pmtud_time_cb`), otherwise state machine will go to blackhole_detected mode.

Reply:
Good catch. I fixed it by always stop the timer when received an ack.